### PR TITLE
Superpixel method now respects the map mask property and allows for arbitrary superpixel dimensions

### PR DESCRIPTION
**Pull request by wafels**
This is a solution to #1536.
Previously, the superpixel map method did not take account of the mask, if it was present.  This PR ensures that the mask is respected when calculating superpixels.  The API has also changed slightly so that a function is passed into superpixel to define how the superpixel is calculated from the original pixels.  This permits other functions (such as numpy.prod) to be passed in by the user without having to change the code.

The existing superpixel method creates a new map from the original map with new pixels formed by summing m by n contiguous original pixels.  If the original map was of size M by N then the old version of the superpixel method required that M/m and N/n be integers.  The new superpixel method lifts this restriction.  Note that if M/m (or N/n) is not an integer, then the new map does not span the same physical extent as the original map.  An offset keyword is provided to allow the user to offset the new map relative to the original map to the extent allowed by requested superpixel size.

### CODE REVIEW

File: sunpy/map/mapbase.py [hunk 1]
```diff
@@ -1089,22 +1089,27 @@ def submap(self, range_a, range_b):
...
-            dimension[1]) pixels of the original map
+            dimension[1]) pixels of the original map.
             Note: the first argument corresponds to the 'x' axis and the second
             argument corresponds to the 'y' axis.
-        method : {'sum' | 'average'}
-            What each superpixel represents compared to the original data
-                * sum - add up the original data
-                * average - average the sum over the number of original pixels
+        offset : tuple
+            Offset from (0,0) used when calculating the
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: missing text?

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,10 @@ Latest
 
 * Added the ability to shift maps to correct for incorrect map location, for example.
 * Bug fix for rhessi summary light curve values.
+* Map superpixel method now respects masks, accepts numpy functions as an
```
Comment (Cadair) [inline CHANGELOG.md]: can you split these up into multiple points?

File: sunpy/map/tests/test_mapbase.py [hunk 3]
```diff
@@ -467,19 +492,23 @@ def test_plot_aia171_nowcsaxes(aia171_test_map):
...
+
+
+@figure_test
+def test_plot_aia171_superpixel_nowcsaxes(aia171_test_map):
+    ax = plt.gca()
+    aia171_test_map.superpixel((9, 7)*u.pix, offset=(4, 4)*u.pix).plot(axes=ax)
+
+
+@figure_test
+def test_plot_masked_aia171_superpixel_nowcsaxes(aia171_test_map_with_mask):
```
Comment (Cadair) [inline sunpy/map/tests/test_mapbase.py]: why only non-wcsaxes plot tests?

File: sunpy/image/tests/test_resample.py [hunk 4]
```diff
@@ -44,8 +45,42 @@ def test_resample_spline():
...
+    im = reshape_image_to_4d_superpixel(aia171_test_map.data, d, o)
+    assert im.shape == (_n(shape[0], o[0], d[0]), d[0],
+                        _n(shape[1], o[1], d[1]), d[1])
+
+    d = (9, 7)
+    o = (4, 4)
+    im = reshape_image_to_4d_superpixel(aia171_test_map.data, d, o)
+    assert im.shape == (_n(shape[0], o[0], d[0]), d[0],
+                        _n(shape[1], o[1], d[1]), d[1])
+
```
Comment (Cadair) [inline sunpy/image/tests/test_resample.py]: style:  whitespace at end of file.

**Comment by Cadair** [comment]
@wafels just a couple of small comments, otherwise looks good.

**Comment by Cadair** [comment]
:+1:

**Comment by Cadair** [comment]
@sunpy/sunpy-developers review please!

**Comment by ayshih** [comment]
:+1:

**Pull request merged**